src/main/java/TcpServer.java

```diff
@@ -0,0 +1,128 @@
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Optional;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class TcpServer implements Closeable {
+  // TODO: make logger instance-specific, prefixing logs w/ config & object info
+  static final Logger logger = Logger.getLogger(TcpServer.class.getName());
+
+  // configuration used to initialize the serverSocket
```

Comment:
For this and a lot of the other fields, the combination of the type and the naming are plenty for me to understand what's going on. I like to use comments only when they really add some new level of understanding. Particularly for cases where I expect that a future maintainer (myself included) would squint and say "huh? why?!"

Most the comments in this class seem to me to be re-iterating the good names that you've already given these fields and methods. A few exceptions are the TODOs (fine by me, provided they don't linger too long in the codebase), and the one on `start()`.

Reply:
Good point, this was just me dallying before starting on the next task while waiting for comments on this one. 😉